be more radical by presenting a non-namespaced view-model in starcoun…

warpech opened the pull request:
…ter-include within a red-dotted outline

Screenshots:

- https://drive.google.com/open?id=1di0Toil0QXRoVvIzj1S1JIXwswA5HZdt
- https://drive.google.com/open?id=1z2LvfWOP8GKI_Mq0p2hTbL7sdxw_6kcJ

This was discussed with @mmnosek and @miyconst.

In the beginning I wanted to be more radical by blurring the view or hiding it completely. This was Konstantin's suggestion to use the outline.

warpech committed:
be more radical by presenting a non-namespaced view-model in starcounter-include within a red-dotted outline

tomalec commented on starcounter-include.html:
```diff
@@ -462,11 +462,15 @@
-`, element);
+                console.warn(`The view ${viewModel.Html} is enclosed in a red-dotted line, because it is incorrectly provided to <starcounter-include> without an app namespace. Your options are: 
+1. Request the blendable Json subtree with Self.GET. Then you can keep to use <starcounter-include> in HTML.
+2. Or, replace <starcounter-include> in HTML with <div><imported-template model="{{viewModel}}" href="{{viewModel.Html}}"></imported-template></div> to insert the Json subtree in a non-blendable fashion. Then you can keep your C# that assigns a Json instance to a parent object without Self.GET.`, viewModel, element);
```
I'd add a note that `declarative-shadow-dom` will not work in second case. As this was a problem for few devs.

tomalec commented on starcounter-include.html:
```diff
@@ -462,11 +462,15 @@
+                element.setAttribute("style", "outline: 4px red dotted; outline-offset: -4px");
+                element.wasWarned = true;
+            }
+            else if(element.wasWarned) {
```
What is this supposed to deliver? Flash the warning every second call to `checkForNoNNamespaced` I found it confusing.

warpech replied:
It removes the border when you replace an non-namespaced view-model with a namespaced view-model. 

Why will it flash every second call?

tomalec replied:
🤦‍♂️ You are right, my bad.

tomalec commented on starcounter-include.html:
```diff
@@ -462,11 +462,15 @@
+                console.warn(`The view ${viewModel.Html} is enclosed in a red-dotted line, because it is incorrectly provided to <starcounter-include> without an app namespace. Your options are: 
+1. Request the blendable Json subtree with Self.GET. Then you can keep to use <starcounter-include> in HTML.
+2. Or, replace <starcounter-include> in HTML with <div><imported-template model="{{viewModel}}" href="{{viewModel.Html}}"></imported-template></div> to insert the Json subtree in a non-blendable fashion. Then you can keep your C# that assigns a Json instance to a parent object without Self.GET.`, viewModel, element);
+                element.setAttribute("style", "outline: 4px red dotted; outline-offset: -4px");
```
I'd use `element.style` not to overwrite upcoming FOUC solution.

warpech commented:
I think/hope that the tests will pass when https://github.com/Starcounter/starcounter-include/pull/111 is merged.

tomalec reviewed (commented):
If you ask me, I'd go for blurring content. But if you really want to make the result look bad and shout at the dev to change it, I'm fine with this border :)

warpech committed:
address @tomalec comments from https://github.com/Starcounter/starcounter-include/pull/114

I can't check getAttribute("style") with equality, because I don't control over how the browser translates styles assigned by ".style" property to the "style" attribute

warpech commented:
Thanks for the review. Blur is too radical, because it makes the page unusable. I'd rather go for annoying than unusable.

I pushed reaction to all comments except https://github.com/Starcounter/starcounter-include/pull/114#discussion_r218102351, to which I wrote an question.

tomalec commented on starcounter-include.html:
```diff
@@ -462,11 +462,16 @@
+                element.wasWarned = true;
+            }
+            else if(element.wasWarned) {
+                element.removeAttribute("style");
```
I'd use `element.style` here as well.

miyconst commented on starcounter-include.html:
```diff
@@ -462,11 +462,16 @@
+                element.style.outlineOffset = "-4px";
+                element.wasWarned = true;
+            }
+            else if(element.wasWarned) {
```
As far as I remember, we agreed to use default Visual Studio formatting for JavaScript and C#. Thus, please make sure that `Ctrl + K + D` does not change the file.

In this particular case it should be:

```js
} else if (element.wasWarned) {
```

miyconst replied:
Should we add a [link](https://docs.starcounter.io/topic-guides/blendable-web-apps/view-model-attaching#attachment-rule-components) to `starcounter.io` for more info regarding `Self.GET` and view attaching?

warpech replied:
Great idea, actually I'll add a link to the blog post that explains the problem.

Regarding the formatting, sorry, I just stick to the formatting that is already in the file.

warpech committed:
reset styles using the .style property

warpech committed:
simplify the warning message and give a link to the blog

warpech commented:
Ready for re-review.

tomalec reviewed (approved):
LGTM, I hope FOUC PR will fix flaky tests

tomalec committed:
Merge remote-tracking branch 'origin/master' into more-radical-error
